### PR TITLE
Space room hierarchies

### PR DESCRIPTION
**Pull request by t3chguy**
image
image
image
image

### CODE REVIEW

File: res/css/structures/_SpaceRoomDirectory.scss [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            > div {
+                color: $secondary-fg-color;
+                font-size: $font-15px;
+                line-height: $font-24px;
+            }
+        }
+    }
+
+    .mx_Dialog_content {
+        // TODO fix scrollbar
```
Comment (turt2live) [inline res/css/structures/_SpaceRoomDirectory.scss:58]: I assume this and the other TODOs are tracked somewhere for your reference

File: src/components/structures/SpaceRoomDirectory.tsx [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+            </React.Fragment>
+        }
+    }
+
+    let url: string;
+    if (space.avatar_url) {
+        url = MatrixClientPeg.get().mxcUrlToHttp(space.avatar_url,
+            Math.floor(24 * window.devicePixelRatio),
+            Math.floor(24 * window.devicePixelRatio),
+            "crop");
```
Comment (turt2live) [inline src/components/structures/SpaceRoomDirectory.tsx:163]: something bothers me about this code style but I don't know what. Not important for merging though - it looks reasonable.

File: src/components/structures/SpaceRoomDirectory.tsx [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+    }, [[], []]) || [[], []];
+
+    // Don't render this subspace if it has no rooms we can show
+    // TODO this is broken - as a space may have subspaces we still need to show
+    // if (!childRooms.length) return null;
+
+    const userId = cli.getUserId();
+
+    const newParents = new Set(parents).add(spaceId);
+    return <React.Fragment>
```
Comment (turt2live) [inline src/components/structures/SpaceRoomDirectory.tsx:367]: shorthand is `<>` btw

**Review by turt2live — APPROVED** [review]
overall seems to do a thing.

File: src/components/structures/SpaceRoomDirectory.tsx (hunk 3 above)
Comment (t3chguy) [reply src/components/structures/SpaceRoomDirectory.tsx:367]: I know, I used to love it but then we didn't support it initially and now am used to this as it also supports `key` unlike the shorthand

File: res/css/structures/_SpaceRoomDirectory.scss (hunk 1 above)
Comment (t3chguy) [reply res/css/structures/_SpaceRoomDirectory.scss:58]: Unfortunately!

File: src/components/structures/SpaceRoomDirectory.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/structures/SpaceRoomDirectory.tsx:163]: image

Could be worse!!

put all params on new lines now

**Pull request merged**